File: java-service/src/main/java/com/analyticore/analyseservice/service/ActivityTracker.java
package com.analyticore.analyseservice.service;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Registro de actividad interna del Servicio Java — solo para el panel / .
 * No almacena estado de negocio (eso vive en PostgreSQL).
 */
public class ActivityTracker {

    private static final String[] INTERNAL_STEPS = {
            "En espera de notificaciones de Python",
            "Recibí POST /api/analyze/{jobId} desde Python",
            "Leyendo el texto del job en PostgreSQL",
            "Actualizando estado → PROCESANDO",
            "Ejecutando análisis de sentimiento",
            "Extrayendo palabras clave del texto",
            "Guardando resultados en PostgreSQL (COMPLETADO)",
            "Análisis finalizado — listo para consulta",
    };

    private static int step = 0;
    private static String message = "En espera — listo para recibir trabajos de Python";
    private static String detail = "";
    private static String mode = "idle";

    private ActivityTracker() {}

    public static void onReceived(String jobId) {
        step = 1;
        message = "Recibí solicitud de análisis desde Python";
        detail = "Job ID: " + jobId;
        mode = "active";
    }

    public static void onReadingDb(String textPreview) {
        step = 2;
        message = "Leyendo texto desde PostgreSQL...";
        detail = "Texto: \"" + textPreview + "\"";
        mode = "active";
    }

    public static void onProcessing() {
        step = 3;
        message = "Actualizando estado del job a PROCESANDO...";
        detail = "UPDATE jobs SET status = 'PROCESANDO'";
        mode = "active";
    }

    public static void onAnalyzingSentiment() {
        step = 4;
        message = "Analizando sentimiento del texto...";
        detail = "Contando palabras positivas y negativas";
        mode = "active";
    }

    public static void onExtractingKeywords() {
        step = 5;
        message = "Extrayendo palabras clave...";
        detail = "Tokenizando y filtrando stopwords";
        mode = "active";
    }

    public static void onSaving(String sentiment, String keywords) {
        step = 6;
        message = "Guardando resultados en PostgreSQL...";
        detail = "Sentimiento: " + sentiment + " | Keywords: " + keywords;
        mode = "active";
    }

    public static void onFinished(String sentiment, double score) {
        step = 7;
        message = "Análisis completado exitosamente";
        detail = "Resultado: " + sentiment + " (score: " + String.format("%.2f", score) + ")";
        mode = "done";
    }

    public static Map<String, Object> getStatus() {
        List<Map<String, String>> steps = new ArrayList<>();
        for (int i = 0; i < INTERNAL_STEPS.length; i++) {
            String status;
            if ("idle".equals(mode) && i == 0) {
                status = "active";
            } else if ("done".equals(mode) || i < step) {
                status = "done";
            } else if (i == step) {
                status = "active";
            } else {
                status = "pending";
            }
            steps.add(Map.of("text", INTERNAL_STEPS[i], "status", status));
        }

        Map<String, Object> result = new HashMap<>();
        result.put("service", "java");
        result.put("message", message);
        result.put("detail", detail);
        result.put("mode", mode);
        result.put("activeStep", step);
        result.put("steps", steps);
        return result;
    }
}

File: python-service/infrastructure/activity_tracker.py
"""
Registro de actividad interna del Servicio Python — solo para el panel / .
No almacena estado de negocio (eso vive en PostgreSQL).
"""
import time
from threading import Lock

_lock = Lock()

# Pasos INTERNOS que ejecuta solo este servicio Python
INTERNAL_STEPS = [
    "En espera de solicitudes del Frontend",
    "Recibí POST /api/jobs desde el Frontend",
    "Validando el texto recibido",
    "Guardando job en PostgreSQL (estado PENDIENTE)",
    "Notificando al Servicio Java vía REST",
    "Java respondió — devolviendo jobId al Frontend",
    "Trabajo entregado correctamente",
]

_state = {
    "step": 0,
    "message": "En espera — listo para recibir textos del Frontend",
    "detail": "",
    "mode": "idle",  # idle | active | done
    "updated_at": time.time(),
}


def _set(step: int, message: str, detail: str = "", mode: str = "active"):
    with _lock:
        _state["step"] = step
        _state["message"] = message
        _state["detail"] = detail
        _state["mode"] = mode
        _state["updated_at"] = time.time()


def on_received(text: str):
    preview = text[:80] + "..." if len(text) > 80 else text
    _set(1, "Recibí texto del Frontend", f'Texto recibido: "{preview}"')


def on_validating():
    _set(2, "Validando formato y longitud del texto...")


def on_saving(job_id: str):
    _set(3, "Guardando en PostgreSQL...", f"Job ID: {job_id} | Estado: PENDIENTE")


def on_calling_java(job_id: str):
    _set(4, "Llamando al Servicio Java...", f"POST /api/analyze/{job_id}")


def on_finished(job_id: str):
    _set(6, "Proceso completado — jobId enviado al Frontend", f"Job ID: {job_id}", mode="done")


def get_status() -> dict:
    with _lock:
        step = _state["step"]
        mode = _state["mode"]
        steps = []
        for i, label in enumerate(INTERNAL_STEPS):
            if mode == "idle" and i == 0:
                status = "active"
            elif mode == "done" or i < step:
                status = "done"
            elif i == step:
                status = "active"
            else:
                status = "pending"
            steps.append({"text": label, "status": status})

        return {
            "service": "python",
            "message": _state["message"],
            "detail": _state["detail"],
            "mode": mode,
            "activeStep": step,
            "steps": steps,
        }
